gammapy/estimators/parameter_estimator.py

```diff
@@ -0,0 +1,183 @@
+        Re-optimize other free model parameters. Default is True.
+    n_scan_values : int
+        Number of values used to scan fit stat profile
+    scan_n_err : float
+        Range to scan in number of parameter error
+    """
+
+    def __init__(
+            self,
+            datasets,
+            sigma=1,
+            sigma_ul=2,
+            reoptimize=True,
+            n_scan_values=30,
+            scan_n_err=3
+    ):
+        self.sigma = sigma
+        self.sigma_ul = sigma_ul
+        self.reoptimize = reoptimize
+        self.n_scan_values = n_scan_values
+        self.scan_n_err = scan_n_err
+
+        self.datasets = self._check_datasets(datasets)
+        self.fit = Fit(datasets)
+        self.fit_result = None
+
+    def __str__(self):
+        s = f"{self.__class__.__name__}:\n"
+        s += str(self.datasets) + "\n"
+        return s
+
+    def _check_datasets(self, datasets):
+        """Check datasets geometry consistency and return Datasets object"""
+        if not isinstance(datasets, Datasets):
+            datasets = Datasets(datasets)
+
+        return datasets
+
+    def _freeze_parameters(self, datasets, parameter):
+        """Freeze all other parameters"""
+        # freeze other parameters
+        for par in datasets.parameters:
+            if par is not parameter:
+                par.frozen = True
+
+    def _compute_scan_values(self, value, value_error, par_min, par_max):
+        """Define parameter value range to be scanned"""
+        min_range = value - self.scan_n_err * value_error
+        if not np.isnan(par_min):
+            min_range = np.maximum(par_min, min_range)
+        max_range = value + self.scan_n_err * value_error
+        if not np.isnan(par_max):
+            max_range = np.minimum(par_max, max_range)
+
+        return np.linspace(min_range, max_range, self.n_scan_values)
+
+    def _find_best_fit(self, parameter):
+        """Find the best fit solution and store results."""
+        fit_result = self.fit.optimize()
+
+        if fit_result.success:
+            value = parameter.value
+        else:
+            value = np.nan
+
+        result = {parameter.name: value,
+                  "stat": fit_result.total_stat,
+                  "success": fit_result.success}
+
+        self.fit_result = fit_result
+        return result
+
+    def _estimate_ts_for_null_value(self, parameter, null_value=1e-150):
+        """Returns the fit statistic value for a given null value of the parameter."""
+        with self.datasets.parameters.restore_values:
+            parameter.value = null_value
+            parameter.frozen = True
+            result = self.fit.optimize()
+            if not result.success or result.message=="Optimization failed.":
+                log.warning("Fit failed for parameter null value, returning NaN. Check input null value.")
+                return np.nan
+            return result.total_stat
+
+    def run(self, parameter, steps, null_value=1e-150):
```

Comment:
Maybe put `steps="all"` as default? And maybe move `null_value` to `__init__()`? I think typically it is not varied but just decided in the beginning..